src/webcomponents/spreadsheet/Spreadsheet.js

```diff
@@ -0,0 +1,165 @@
+import React, { useRef, useEffect, useState } from "react";
+import jspreadsheet from "jspreadsheet-ce";
+import data from '../../data/data.json';
+
+import "../../../node_modules/jspreadsheet-ce/dist/jexcel.css";
+
+export default function App(props) {
+  //const [sJson,setSJson] = useState('')
+  let json;
+  const changed = async()=> {
+    //const cellName = jexcel.getColumnNameFromId([x,y]);
+    json = await jRef.current.jexcel.getJson();
+    console.log('on',json);
+  }
+
+  const onafterchanged = async()=> {
+    json = await jRef.current.jexcel.getJson();
+    console.log('after',jRef.current.jexcel.getJson());
+  }
+
+  const oninserted = async ()=>{
+    json = await jRef.current.jexcel.getJson();
+    console.log('in',jRef.current.jexcel.getJson());
+  }
+  const ondeleted =async ()=> {
+    json = await jRef.current.jexcel.getJson();
+    console.log('del',jRef.current.jexcel.getJson());
+  }
+  
+  
+const datas = JSON.parse(props.languages);
+let dataArray = [];
+let keys = [];
+let columns = [];
+let cols = []
+let rows = [];
+let rowData = [];
+let c = [{ type:'text',title:"messageID", width:150 }];
+if(datas){
+    for(const d in datas){
+        c.push({ type:'text',title:`${d}`, width:150,data:'nice' });
+        dataArray.push(datas[d])
+    }
+}
+if(dataArray){
+    for (const d in dataArray){
+        columns.push(dataArray[d])
+    }
+    for (const d in dataArray[0]){
+        // console.log('dada',d);
+        keys.push(d)
+    }
+}
+//console.log(Object.keys(dataArray[0]));
+// console.log('keys',keys);
+if(columns){
```

Comment:
Extract the code that transforms from languages to jspreadsheet data structure to a separate typescript file. Each loop should be wrapped as a function to improve readability (the name of the functions helps understanding of the code).